api/src/retry.rs

```diff
@@ -0,0 +1,204 @@
+use reqwest::{blocking::Response, Result};
+use std::sync::atomic::{AtomicBool, Ordering::Relaxed};
+use std::thread::sleep;
+use std::time::Duration;
+
+/// Strategy to use if retrying .
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum RetryStrategy {
+    /// The first request by the client will not be retried, but subsequent requests will.
+    /// This allows fast failure if the client cannot reach the API endpoint at all, but
+    /// helps to mitigate failure in long-running operations spanning multiple requests.
+    Automatic,
+    /// Always attempt to retry requests.
+    Always,
+}
+
+/// Configuration for the Reinfer client if retrying timeouts.
+#[derive(Clone, Debug, PartialEq)]
+pub struct RetryConfig {
+    /// Strategy for when to retry after a timeout
+    pub strategy: RetryStrategy,
+    /// Maximum number of retries to attempt.
+    pub max_retry_count: u8,
+    /// Amount of time to wait for first retry.
+    pub base_wait: Duration,
+    /// Amount of time to scale retry waits. The wait before retry N is an exponential backoff
+    /// using the formula `wait = retry_wait * (backoff_factor * N)`.
+    pub backoff_factor: f64,
+}
+
+#[derive(Debug)]
+pub(crate) struct Retrier {
+    config: RetryConfig,
+    is_first_request: AtomicBool,
+}
+
+impl Retrier {
+    pub fn new(config: RetryConfig) -> Self {
+        Self {
+            config,
+            is_first_request: AtomicBool::new(true),
+        }
+    }
+
+    pub fn with_retries(&self, f: impl Fn() -> Result<Response>) -> Result<Response> {
+        if self.is_first_request.swap(false, Relaxed)
```

Comment:
Why `Relaxed`?  Seems like a pretty insignificant cost to use `SeqCst` here and saves the reader having to reason about the correctness and potentially prevents subtle bugs. Happy to have my mind changed, but my experience with C++ atomics means my stance is that not using `SeqCst` is and optimisation, and therefore has to be justified by performance numbers.

Reply:
My thinking here was that it's not immediately obvious anyway what behaviour is expected if two threads were to launch "the first request" simultaneously. Probably from a user's perspective it's fine for both of them to fail immediately without retry. This is, after all, just trying to make an interface more friendly rather than doing some work that needs to have guaranteed sychronisation.

So the `Relaxed` was because I literally didn't think it mattered in this case. (I considered `Cell` instead of `AtomicBool` but didn't feel I needed to force this type into being `!Sync`.) I completely agree with you that the other 99% of the time `Relaxed` is scary, so happy to use `SeqCst` here just to prevent people getting into bad habits.